src/server/bg_services/bucket_logs_upload.js

```diff
@@ -178,4 +187,76 @@ class BucketLogUploader {
+    const log = new PersistentLogger(config.PERSISTENT_BUCKET_LOG_DIR, config.PERSISTENT_BUCKET_LOG_NS, { locking: 'EXCLUSIVE' });
+    try {
+        await log.process(async file => _upload_to_targets(fs_context, s3_connection, file, bucket_to_owner_keys_func));
+    } catch (err) {
+        dbg.error('processing log file failed', log.file);
+        throw err;
+    } finally {
+        await log.close();
+    }
+}
+
+/**
+ * This function gets a persistent log file, will go over it's entries one by one,
+ * and will upload the entry to the target_bucket using the provided s3 connection
+ * in order to know which user to use to upload to each bucket we will need to provide bucket_to_owner_keys_func
+ * @param {nb.NativeFSContext} fs_context
+ * @param {AWS.S3} s3_connection
+ * @param {string} log_file
+ * @param {function} bucket_to_owner_keys_func
+ * @returns {Promise<Boolean>}
+ */
+async function _upload_to_targets(fs_context, s3_connection, log_file, bucket_to_owner_keys_func) {
+    const bucket_streams = {};
+    const promises = [];
+    try {
+        const file = new LogFile(fs_context, log_file);
+        dbg.log1('uploading file to target buckets', log_file);
+        await file.collect_and_process(async entry => {
+            const log_entry = JSON.parse(entry);
+            nsfs_schema_utils.validate_log_schema(log_entry);
+            const target_bucket = log_entry.log_bucket;
+            const log_prefix = log_entry.log_prefix;
+            const source_bucket = log_entry.source_bucket;
+            if (!bucket_streams[source_bucket]) {
+                const date = new Date();
+                const upload_stream = new stream.PassThrough();
+                const access_keys = await bucket_to_owner_keys_func(target_bucket);
+                s3_connection.config.credentials.accessKeyId = access_keys[0].access_key;
+                s3_connection.config.credentials.secretAccessKey = access_keys[0].secret_key;
+                const sha = crypto.createHash('sha512').update(target_bucket + date.getTime()).digest('hex');
+                promises.push(sem.surround(() => P.retry({
+                    attempts: 3,
+                    delay_ms: 1000,
+                    func: () => s3_connection.upload({
+                        Bucket: target_bucket,
+                        Key: `${log_prefix}${format_aws_date(date)}-${sha.slice(0, 16).toUpperCase()}`,
+                        Body: upload_stream,
+                    }).promise()
+                })));
+                bucket_streams[source_bucket] = upload_stream;
+            }
+            dbg.log2(`uploading entry: ${entry} to target bucket: ${target_bucket}`);
+            bucket_streams[source_bucket].write(entry + '\n');
+        });
+        Object.values(bucket_streams).forEach(st => st.end());
+        await Promise.all(promises);
+    } catch (error) {
+        dbg.error('unexpected error in upload to bucket:', error, 'for:', log_file);
```

Comment:
do we want to have a cleanup of incomplete logs in case of an error?

Reply:
As we seen together no cleanup is needed